### PR TITLE
[20044] Upgrade to Rails 4.2.x

### PR DESCRIPTION
**Pull request by myabc**
:warning: **Please review, merge opf/openproject-translations#29 first**, adjusting branches in `Gemfile` as appropriate.
:warning: ~~**Includes commits from #3382. Please review, merge first.**~~

https://community.openproject.org/work_packages/20044

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://myabc-feature-rails4-2.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://myabc-feature-rails4-2.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by basiszwo** [comment]
- [ ] remove deprecation warning about `sanitize_sql_hash_for_conditions` by upgrading act_as_list. see https://github.com/swanandp/acts_as_list/pull/140

File: app/controllers/account_controller.rb [hunk 1]
```diff
@@ -100,7 +100,7 @@ def lost_password
         # create a new token for password recovery
         token = Token.new(user: user, action: 'recovery')
         if token.save
-          UserMailer.password_lost(token).deliver
+          UserMailer.password_lost(token).deliver_now
```
Comment (ulferts) [inline app/controllers/account_controller.rb:103]: I realize, this is the 1:1 translation but can't we use `deliver_later` instead @basiszwo, @myabc? Would this use our readily available delayed_job then?

**Comment by ulferts** [comment]
@myabc would you be so kind to rebase this on the current dev now that #3382 is merged?

File: app/models/project.rb [hunk 2]
```diff
@@ -860,13 +860,18 @@ def add_work_package(attributes = {})
     WorkPackage.new do |i|
       i.project = self
 
-      type_attribute = attributes.delete(:type) || attributes.delete(:type_id)
-
-      i.type = if type_attribute
-                 project.types.find(type_attribute)
+      type    = attributes.delete(:type)
+      type_id = if type && type.respond_to?(:id)
```
Comment (myabc) [inline app/models/project.rb:865]: @ulferts am I being overly defensive here?

Comment (basiszwo) [inline app/models/project.rb:865]: In which case is it possible that type does not respond to `:id` ?

File: app/controllers/account_controller.rb (hunk 1 above)
Comment (myabc) [inline app/controllers/account_controller.rb:103]: @ulferts sounds like a good idea, but perhaps better as a separate PR?

File: app/models/project.rb (hunk 2 above)
Comment (myabc) [inline app/models/project.rb:865]: @basiszwo I'm not that familiar with the `WorkPackage#add_work_package` API - I think this is a Redmine remnant. It's conceivable though that a user passes in an attributes `Hash` with a `type: nil`, e.g. `type: Type.find(by_name: 'unknown')`

**Comment by myabc** [comment]
@ulferts @basiszwo @NobodysNightmare @oliverguenther we're getting closer to green CI. In order to keep things on track, I'd really appreciate some help/guidance in two areas:
- [x] ~~`Setting` callbacks no longer seem to work (at least in specs). This also causes `AvatarHelper` specs to fail.~~ https://travis-ci.org/opf/openproject/jobs/80118468#L2578
- [x] `Activity`: https://travis-ci.org/opf/openproject/jobs/80118468#L2653 I also marked a legacy spec as pending. 97b24b94533ec9577867590c6c2fd251476f2501 - @ulferts does this look like a permissions issue?

**Comment by oliverguenther** [comment]
> Setting callbacks no longer seem to work (at least in specs). 

This appears to be due to the symbol that is passed to `instrument`, which appears to now be used directly rather than stringified in the AS instrumenter. Changing the call to use a string at all times seems to fix it.

/Edit: Ah. I see you just addressed this yourself. Never mind then.

I've opened https://github.com/myabc/openproject/pull/8 for it, going to push more when I find it.

**Comment by myabc** [comment]
> /Edit: Ah. I see you just addressed this yourself. Never mind then.

@oliverguenther sorry for the duplicated effort (you're too fast, or I'm too slow :wink:).

**Comment by ulferts** [comment]
> @ulferts does this look like a permissions issue?

@myabc I'll take a look at it once I am done with my current review.

**Comment by ulferts** [comment]
@myabc I assume to have fixed the [failing activity spec](https://github.com/opf/openproject/pull/3421#issuecomment-139910923)  in myabc#9. There seems to be a change in the arel behaviour when you pass nested arrays to `in`. It apparently used to be flattened but now the arel visitor will transform it each array to NULL

File: features/types/types_adminstration.feature [hunk 3]
```diff
@@ -90,4 +90,4 @@ Feature: Type Administration
   Scenario: A standard type cannot be deleted
      When I go to the global index page of types
      Then "Standard" should be the last element in the list
```
Comment (myabc) [inline features/types/types_adminstration.feature:92]: Note: `first element`/`last element` steps are actually broken (they don't check position in the list). Should be fixed here and in other places, but out-of-scope for this PR probably.

**Comment by myabc** [comment]
@ulferts @basiszwo @NobodysNightmare @oliverguenther core is now ready for review. Note: I'm seeing some flickering with `user_mailer_spec`. Not sure if this also affects Rails 4.1/`dev`.

**Comment by myabc** [comment]
@oliverguenther any idea why the Repository features, in particular, might be flickering? https://travis-ci.org/opf/openproject/jobs/80581422

**Comment by oliverguenther** [comment]
@myabc We are seeing this independent of the rails. 
I've tried to fix with this one on the CI several times, and I remember @ulferts having a go at it, too. I'll try to refactor the functionality, as it includes a full JS redirect and there seems to be some bug finding elements afterwards.

**Comment by myabc** [comment]
@oliverguenther ok, thanks for clarification!

**Comment by myabc** [comment]
@ulferts @basiszwo @NobodysNightmare @oliverguenther core + plugins are now ready for review (unfortunately I wasn't able to get a completely green build on Travis)

**Comment by oliverguenther** [comment]
Changes in this PR look good to me, going to let the various configs run through.

**Comment by ulferts** [comment]
Removing the WIP label, then. I hope to get a look at it today @myabc.

**Comment by oliverguenther** [comment]
/Edit: Hmm,.. is working now. I blame `db:test:prepare`, although I ran this on a clean database...

**Comment by myabc** [comment]
@oliverguenther what was the issue?

**Comment by oliverguenther** [comment]
@myabc I was seeing numerous primary key errors on the myproject config upon saving / updating work packages, both in specs and on the UI.

```
ActiveRecord::RecordNotUnique (PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "work_package_journals_pkey"
DETAIL:  Key (id)=(2) already exists.
: INSERT INTO "work_package_journals" ("id", "type_id", "project_id", "subject", "description", "status_id", "priority_id", "author_id", "journal_id") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9) RETURNING "id"):
  app/models/journal.rb:130:in `save_data'
  lib/plugins/acts_as_journalized/lib/redmine/acts/journalized/save_hooks.rb:95:in `save_journal_with_retry'
  lib/plugins/acts_as_journalized/lib/redmine/acts/journalized/save_hooks.rb:73:in `block in save_journals'
  lib/plugins/acts_as_journalized/lib/redmine/acts/journalized/save_hooks.rb:72:in `each'
  lib/plugins/acts_as_journalized/lib/redmine/acts/journalized/save_hooks.rb:72:in `save_journals'
  app/services/update_work_package_service.rb:46:in `save'
  lib/api/v3/work_packages/work_packages_api.rb:80:in `block (3 levels) in <class:WorkPackagesAPI>'
  app/middleware/params_parser_with_exclusion.rb:38:in `call'
```

**Comment by oliverguenther** [comment]
Now I'm seeing the two following expected errors:

~~`can't modify frozen ActiveSupport::HashWithIndifferentAccess`
`silent_list` uses `changed_attributes[key] = foo`, which appears to be invalid in Rails 4.2:~~

Darn, you already fixed this and I've not checked out the 4.2 branch on silent_list. Never mind!

```
     Failure/Error: let(:type) { FactoryGirl.create(:type_standard) }
     ActiveRecord::RecordInvalid:
       Validation failed: Name has already been taken.
```

**Comment by ulferts** [comment]
This s looking good to merge. Thanks for the tremendous job @myabc 

Will merge once the configurations are healthy again. They got broken by some other merges and I want to separate potential test breakages.

**Comment by myabc** [comment]
:horse:

**Pull request merged**